### PR TITLE
mzML Files changed to comply with mzML format specifications

### PR DESCRIPTION
**Pull request by dippman**
mzML Files changed to comply with mzML format specifications. Allows use of third party software to open and view mzML files generated by mzLib.

Added Chromatogram generation

### CODE REVIEW

File: MzML/Mzml.cs [hunk 1]
```diff
@@ -363,11 +365,11 @@ private static int GetOneBasedPrecursorScanNumber(Generated.mzMLType _mzMLConnec
             do
             {
                 oneBasedSpectrumNumber--;
-            } while (!precursorID.Equals(_mzMLConnection.run.spectrumList.spectrum[oneBasedSpectrumNumber - 1].id));
+            } while (!precursorID.Equals(_mzMLConnection.run.spectrumList.spectrum[oneBasedSpectrumNumber].id));
```
Comment (stefanks) [inline MzML/Mzml.cs]: What if there are two ms1 scans, where the first one is the actual precursor? Could you instead check if the scan  id matches?

File: MzML/MzmlMethods.cs [hunk 2]
```diff
@@ -98,37 +109,211 @@ public static void CreateAndWriteMyMzmlWithCalibratedSpectra(IMsDataFile<IMsData
                 version = "1",
                 cvParam = new Generated.CVParamType[1]
             };
+            mzML.softwareList.software[1] = new Generated.SoftwareType()
+            {
+                id = "pwiz",
```
Comment (stefanks) [inline MzML/MzmlMethods.cs]: Why do we need to mention pwiz and xcalibur in our mzml files?

File: MzML/MzmlMethods.cs [hunk 3]
```diff
@@ -140,34 +325,47 @@ public static void CreateAndWriteMyMzmlWithCalibratedSpectra(IMsDataFile<IMsData
...
                 {
-                    defaultArrayLength = myMsDataFile.GetOneBasedScan(i).MassSpectrum.Size,
-
-                    index = i.ToString(CultureInfo.InvariantCulture),
-                    id = myMsDataFile.GetOneBasedScan(i).OneBasedScanNumber.ToString(),
-
-                    cvParam = new Generated.CVParamType[8]
+                    defaultArrayLength = myMsDataFile.GetOneBasedScan(i).MassSpectrum.YArray.Length,
+                    index = (i - 1).ToString(CultureInfo.InvariantCulture),
+                    id = "controllerType=0 controllerNumber=1 " + "scan=" + (myMsDataFile.GetOneBasedScan(i).OneBasedScanNumber).ToString(),
```
Comment (stefanks) [inline MzML/MzmlMethods.cs]: Why does it need to be in this format?

File: MzML/MzmlMethods.cs [hunk 4]
```diff
@@ -406,109 +656,117 @@ public static void CreateAndWriteMyMzmlWithCalibratedSpectra(IMsDataFile<IMsData
...
-                    };
-                    mzML.run.spectrumList.spectrum[i - 1].binaryDataArrayList.binaryDataArray[4].userParam = new Generated.UserParamType[1];
-                    mzML.run.spectrumList.spectrum[i - 1].binaryDataArrayList.binaryDataArray[4].userParam[0] = new Generated.UserParamType()
-                    {
-                        name = "kelleherCustomType",
-                        value = "noise intensity"
-                    };
-                }
+                //Following seems to be uneeded. keep code for now incase.
+                /*
```
Comment (stefanks) [inline MzML/MzmlMethods.cs]: At least some users wanted this. Can it be enabled while still working with other software?

File: MzML/MzmlMethods.cs (hunk 2 above)
Comment (dippman) [inline MzML/MzmlMethods.cs]: Woops, forgot to remove those we don't!

File: MzML/MzmlMethods.cs (hunk 3 above)
Comment (dippman) [inline MzML/MzmlMethods.cs]: This format was taken from other mzML files that all seemed to specify the controllerType and Number in every spectrum, although it is not needed. The minimum required would look something like "scan=(index)"

Comment (stefanks) [inline MzML/MzmlMethods.cs]: Yeah, let's do the minimal required, as long as it doesn't break other software

File: MzML/MzmlMethods.cs [hunk 5]
```diff
@@ -61,34 +61,45 @@ public static void CreateAndWriteMyMzmlWithCalibratedSpectra(IMsDataFile<IMsData
         {
             var mzML = new Generated.mzMLType()
             {
-                version = "1",
-                cvList = new Generated.CVListType()
+                version = "1.18.2",
```
Comment (stefanks) [inline MzML/MzmlMethods.cs]: change please

File: MzML/MzmlMethods.cs [hunk 6]
```diff
@@ -61,34 +61,45 @@ public static void CreateAndWriteMyMzmlWithCalibratedSpectra(IMsDataFile<IMsData
...
-                accession = "MS:1000580" // MSn Data
+                accession = "MS:1000580", // MSn Data
+                name = "MSn spectrum",
+                cvRef = "MS"
             };
             mzML.softwareList = new Generated.SoftwareListType()
             {
                 count = "1",
-                software = new Generated.SoftwareType[1]
+                software = new Generated.SoftwareType[3]
```
Comment (stefanks) [inline MzML/MzmlMethods.cs]: back to 1

File: MzML/MzmlMethods.cs [hunk 7]
```diff
@@ -98,37 +109,185 @@ public static void CreateAndWriteMyMzmlWithCalibratedSpectra(IMsDataFile<IMsData
...
+            double[] times = new double[myMsDataFile.NumSpectra];
+            double[] intensities = new double[myMsDataFile.NumSpectra];
+
+            for (int i = 1; i <= myMsDataFile.NumSpectra; i++)
+            {
+                times[i - 1] = myMsDataFile.GetOneBasedScan(i).RetentionTime;
+                intensities[i - 1] = myMsDataFile.GetOneBasedScan(i).MassSpectrum.SumOfAllY;
+            }
+
+            byte[] test = { (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A', (byte)'A' };
```
Comment (stefanks) [inline MzML/MzmlMethods.cs]: get rid please

File: MzML/MzmlMethods.cs [hunk 8]
```diff
@@ -140,34 +299,54 @@ public static void CreateAndWriteMyMzmlWithCalibratedSpectra(IMsDataFile<IMsData
...
-                    mzML.run.spectrumList.spectrum[i - 1].cvParam[0].accession = "MS:1000580";
-
+                    mzML.run.spectrumList.spectrum[i - 1].cvParam[0] = new Generated.CVParamType
+                    {
+                        accession = "MS:1000580",
+                        cvRef = "MS",
+                        name = "MSn spectrum",
+                        value = ""
+                    };
+                    string precursorID = mzML.run.spectrumList.spectrum[i - 1].id;
```
Comment (stefanks) [inline MzML/MzmlMethods.cs]: needs to be the precursor scan id, not id of the current scan. 
Used to be 
'
string precursorID = scanWithPrecursor.OneBasedPrecursorScanNumber.ToString();
'

Comment (stefanks) [inline MzML/MzmlMethods.cs]: precursorID = "scan="+scanWithPrecursor.OneBasedPrecursorScanNumber

**Pull request merged**